### PR TITLE
VL-20_Move-UStepper-logic-to-UProgress-component_Dmytro_Holdobin

### PR DESCRIPTION
**Pull request by Explicit12**
-Moved UStepper component logic to UProgress component

> https://ilevel.atlassian.net/browse/VL-20

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -68,5 +68,6 @@
   "bugs": {
     "url": "https://github.com/vuelessjs/vueless/issues"
   },
-  "web-types": "./web-types.json"
+  "web-types": "./web-types.json",
+  "packageManager": "npm@10.7.0^"
```
Comment (itsJohnnyGrid) [inline package.json:72]: please don't specify package manager. Dev's can use whatever they want.

File: src/ui.navigation-stepper/index.vue [hunk 2]
```diff
@@ -106,6 +106,15 @@ const props = defineProps({
   },
 });
 
+// TODO: Replace with UProgress in all our project and delete this component
+console.warn(
+  `
+    UStepper component is deprecated. 
+    Use UProgress with stepper type props instead. 
+    This component will be removed in the future.
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-stepper/index.vue:114]: you can remove it now.

File: src/ui.navigation-progress/components/StepperProgress.vue [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  dataCy: {
+    type: String,
+    default: "",
+  },
+});
+
+const { stepperRingAttrs, stepperCountAttrs, stepperGradientAttrs, stepperSvgAttrs } =
+  useAttrs(props);
+
+const setStepperColor = computed(() => {
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/components/StepperProgress.vue:80]: why computed called as a method?

File: src/ui.navigation-progress/components/StepperProgress.vue [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+<script setup>
+import { computed } from "vue";
+import colors from "tailwindcss/colors";
+
+import { GRAY_COLORS } from "../../preset.tailwind/constants";
+
+import { useAttrs } from "../composables/attrs.composable";
+
+const props = defineProps({
+  title: {
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/components/StepperProgress.vue:46]: this prop doesn't use in the component

File: src/ui.navigation-progress/components/StepperProgress.vue [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+const setStepperColor = computed(() => {
+  return colors[props.color]
+    ? colors[props.color][500]
+    : GRAY_COLORS.includes(props.color)
+      ? colors[props.color][900]
+      : colors.zinc[900];
+});
+
+const fillPercentage = computed(() => {
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/components/StepperProgress.vue:88]: As I remember we already have smth similar inside `UProgress`

File: src/ui.navigation-progress/components/StepperProgress.vue [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    </defs>
+
+    <circle
+      cx="20"
+      cy="20"
+      r="15.91549430918954"
+      fill="transparent"
+      stroke-width="4"
+      v-bind="stepperRingAttrs"
+    ></circle>
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/components/StepperProgress.vue:17]: use self closing tag ` />`

File: src/ui.navigation-progress/composables/attrs.composable.js [hunk 7]
```diff
@@ -6,7 +6,19 @@ import { computed } from "vue";
 
 export function useAttrs(props) {
   const { config, getAttrs, getColor, setColor } = useUI(defaultConfig, () => props.config);
-  const { progress, indicator, step } = config.value;
+  const { progress, indicator, step, wrapper, stepper } = config.value;
+
+  const cvaStepper = cva({
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/composables/attrs.composable.js:11]: migrate the composable to the new architecture

File: src/ui.navigation-progress/configs/default.config.js [hunk 8]
```diff
@@ -1,5 +1,32 @@
...
+        xs: "h-8 w-8",
+        sm: "h-10 w-10",
+        md: "h-12 w-12",
+        lg: "h-14 w-14",
+        xl: "h-16 w-16",
+        "2xl": "h-20 w-20",
+      },
+    },
+  },
+  stepperRing: "stroke-current text-gray-100",
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/configs/default.config.js:25]: `stepperCircle` is better

File: src/ui.navigation-progress/components/StepperProgress.vue [hunk 9]
```diff
@@ -0,0 +1,94 @@
+<template>
+  <svg width="100%" height="100%" viewBox="0 0 40 40" v-bind="stepperSvgAttrs">
+    <defs>
+      <linearGradient id="gradient" x1="0%" y1="0%" x2="100%" y2="0%" v-bind="stepperGradientAttrs">
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/components/StepperProgress.vue:4]: how we can use this?

File: src/ui.navigation-progress/configs/default.config.js [hunk 10]
```diff
@@ -65,6 +94,7 @@ export default /*tw*/ {
   defaultVariants: {
     color: "brand",
     size: "md",
+    type: "progress",
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/configs/default.config.js:97]: type >>> variant

File: src/ui.navigation-progress/constants/index.js [hunk 11]
```diff
@@ -3,3 +3,8 @@
  `defineProps` is referencing locally declared variables. (vue/valid-define-props)
  */
 export const UProgress = "UProgress";
+
+export const TYPES = {
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/constants/index.js:7]: TYPES >>> VARIANT (single for the object)

File: src/ui.navigation-progress/index.vue [hunk 12]
```diff
@@ -72,16 +90,49 @@ const props = defineProps({
     default: UIService.get(defaultConfig, UProgress).default.color,
   },
 
+  /**
+   * Stepper type title.
+   */
+  stepperTitle: {
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/index.vue:96]: do the same as we have in progress variant (each step may have its own title from the array of steps)

the prop no needed.

File: src/ui.navigation-progress/components/StepperProgress.vue (hunk 9 above)
Comment (Explicit12) [reply src/ui.navigation-progress/components/StepperProgress.vue:4]: That's colored progress circle

File: src/ui.navigation-progress/components/StepperProgress.vue (hunk 3 above)
Comment (itsJohnnyGrid) [reply src/ui.navigation-progress/components/StepperProgress.vue]: why its not a computed now? How about reactitvity?

File: src/ui.navigation-progress/configs/default.config.js (hunk 8 above)
Comment (itsJohnnyGrid) [reply src/ui.navigation-progress/configs/default.config.js:25]: not fixed

File: src/ui.navigation-progress/index.vue [hunk 13]
```diff
@@ -21,6 +23,17 @@
         </slot>
       </template>
     </div>
+
+    <template v-if="isVariant.stepper">
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/index.vue:27]: template tag no needs here

File: src/ui.navigation-progress/index.vue [hunk 14]
```diff
@@ -93,7 +126,21 @@ const realMax = computed(() => {
...
+
+const progressPercent = computed(() => {
+  const maxPercent = 100;
+  const currentPercent = (props.value / realMax.value) * maxPercent;
+
+  if (isVariant.value.progress) {
+    return `${currentPercent}%`;
+  }
+
+  return `${currentPercent} ${maxPercent}`;
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/index.vue:142]: this is not clear for me, need to discuss

Comment (Explicit12) [reply src/ui.navigation-progress/index.vue:142]: It's applied to `<circle>` stroke-dasharray attribute when variant is Stepper, so we need this data format

File: src/ui.navigation-progress/composables/attrs.composable.js [hunk 15]
```diff
@@ -5,72 +5,49 @@ import defaultConfig from "../configs/default.config";
...
+    let classes = "";
+    let value = config.value[key];
 
-  const cvaStep = cva({
-    base: step.base,
-    variants: step.variants,
-    compoundVariants: step.compoundVariants,
-  });
+    const isVariants = value.variants || value.compoundVariants;
+    const excludeKeys = ["progress", "indicator", "step"];
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/composables/attrs.composable.js:21]: why this need?

File: src/ui.navigation-progress/composables/attrs.composable.js [hunk 16]
```diff
@@ -5,72 +5,49 @@ import defaultConfig from "../configs/default.config";
...
-        color: getColor(props.color),
-      }),
-      props.color,
-    ),
-  );
+    if (isVariants && excludeKeys.includes(key)) {
+      const getCVA = cva(value);
+
+      classes = computed(() =>
+        setColor(getCVA({ ...props, color: getColor(props.color) }), props.color),
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/composables/attrs.composable.js:33]: you should apply this staff to getCVA directly to allow to set color to any of config keys when user needs.

File: src/ui.navigation-progress/composables/attrs.composable.js [hunk 17]
```diff
@@ -5,72 +5,49 @@ import defaultConfig from "../configs/default.config";
...
+      );
+    }
 
-  const wrapperAttrs = getAttrs("wrapper");
-  const indicatorAttrs = getAttrs("indicator", { classes: indicatorClasses });
-  const progressAttrs = getAttrs("progress", { classes: progressClasses });
-  const stepAttrsRaw = getAttrs("step", { classes: stepClasses });
+    if (key === "step") {
+      const stepAttrsRaw = getAttrs("step", { classes });
```
Comment (itsJohnnyGrid) [inline src/ui.navigation-progress/composables/attrs.composable.js:38]: Raw suffix no needed anymore.

File: src/ui.navigation-progress/index.vue (hunk 14 above)
Comment (itsJohnnyGrid) [reply src/ui.navigation-progress/index.vue:141]: its better to divide this computed into 2 computed. and calculate `${currentPercent} ${maxPercent}` inside the component where it need to be used.

File: src/ui.navigation-progress/composables/attrs.composable.js (hunk 15 above)
Comment (Explicit12) [reply src/ui.navigation-progress/composables/attrs.composable.js:21]: The idea was to separate keys which needed color props and apply setColor function only to them in favor of better optimization. Taking in attention your [next comment](https://github.com/vuelessjs/vueless/pull/64#discussion_r1670622468)  we don't actually need it.

**Review by itsJohnnyGrid — APPROVED** [review]

**Pull request merged**